### PR TITLE
feat: vectorize merge in panelplot path

### PR DESCRIPTION
**Pull request by dannygoldstein**
#### Summary
This PR introduces the capability to perform vectorized scalar,vector and vector,scalar merges in weave. Before only vector,vector merges were supported. The scalar,vector case is often hit in the PanelPlot path. 

#### Changes
1. Added `_ensure_awl_for_merge` in `weave/ops_arrow/dict.py` to convert scalar dicts to ArrowWeaveLists before merging.
2. Introduced `preprocess_merge` op to ensure that inputs for merge are ArrowWeaveLists.
3. Added `_vectorize_merge_special_case` in `weave/ops_arrow/vectorize.py` to use the `preprocess_merge`.
4. Updated type handling logic in `weave/ops_primitives/_dict_utils.py`.
5. Added relevant tests for new merge behaviors in `weave/tests/test_arrow_vectorizer.py`.

#### Testing
New unit tests have been added to validate the functionality. 

#### Impact
This PR should improve the robustness of merging operations in `weave` when using ArrowWeaveList and Python dicts interchangeably. It should result in performance improvements, specifically for panelplots with many rows.

### CODE REVIEW

File: weave/ops_primitives/_dict_utils.py [hunk 1]
```diff
@@ -32,8 +32,15 @@ class MergeInputTypes(typing.TypedDict):
 def typeddict_merge_output_type(
     input_types: MergeInputTypes,
 ) -> typing.Union[types.TypedDict, types.UnionType, types.NoneType, types.UnknownType]:
-    self = input_types["self"]
-    other = input_types["other"]
+    self: types.Type = input_types["self"]
```
Comment (tssweeney) [inline weave/ops_primitives/_dict_utils.py:35]: do you have a sense as to why we need to do Const unwrapping here, but not in all output type cases?

Comment (dannygoldstein) [reply weave/ops_primitives/_dict_utils.py:35]: because of L45 (RHS)

File: weave/ops_arrow/dict.py [hunk 2]
```diff
@@ -111,6 +111,47 @@ def pick(self, key):
...
+        )
+    return ensure_target
+
+
+MaybeTypedDictType = types.optional(types.TypedDict({}))
+AWLMaybeTypedDictType = ArrowWeaveListType(MaybeTypedDictType)
+
+
+@op(
+    name="ArrowWeaveList-_preprocessMerge",
```
Comment (tssweeney) [inline weave/ops_arrow/dict.py:134]: do we use this underscore convention elsewhere? Is this just to make it extra hidden?

File: weave/ops_arrow/dict.py [hunk 3]
```diff
@@ -111,6 +111,47 @@ def pick(self, key):
...
+    input_type={
+        "self": types.union(AWLMaybeTypedDictType, MaybeTypedDictType),
+        "other": (
+            lambda input_types: AWLMaybeTypedDictType
```
Comment (tssweeney) [inline weave/ops_arrow/dict.py:138]: nice

File: weave/ops_arrow/dict.py (hunk 2 above)
Comment (dannygoldstein) [reply weave/ops_arrow/dict.py:134]: exactly. its just to dissuade anyone from ever trying to use it

File: weave/ops_arrow/dict.py [hunk 4]
```diff
@@ -111,6 +111,47 @@ def pick(self, key):
...
 
+def _ensure_awl_for_merge(
+    ensure_target: typing.Union[dict, ArrowWeaveList],
+    compare_target: typing.Union[dict, ArrowWeaveList],
+) -> ArrowWeaveList:
+    if not isinstance(ensure_target, ArrowWeaveList):
+        assert isinstance(compare_target, ArrowWeaveList)
+        utility_awl = convert.to_arrow([ensure_target])
+        return ArrowWeaveList(
+            pa.repeat(ensure_target, len(compare_target)),
```
Comment (tssweeney) [inline weave/ops_arrow/dict.py:122]: is `repeat` fast?

**Review by tssweeney — APPROVED** [review]
Looks really clean! Nice work

File: weave/ops_arrow/dict.py (hunk 4 above)
Comment (dannygoldstein) [reply weave/ops_arrow/dict.py:122]: Yep. We use it in other places in the codebase to achieve something similar

**Pull request merged**